demos/gplates_global/forward.py

```diff
@@ -0,0 +1,370 @@
+
+    time = timestepping_history.get("time")[-1]
+
+    # Now perform the time loop:
+    while time < ndtime_now:
+        # Update surface velocities
+        gplates_velocities.update_plate_reconstruction(time)
+
+        # Solve Stokes system:
+        stokes_solver.solve()
+
+        # Adapt time step
+        if num_timestep != 0:
+            dt = t_adapt.update_timestep()
+        else:
+            dt = float(delta_t)
+        time += dt
+
+        # Make sure we are not going past present day
+        if ndtime_now - time < float(dt):
+            dt = ndtime_now - time
+
+        # Temperature system:
+        energy_solver.solve()
+
+        time += float(delta_t)
+        num_timestep += 1
+
+        # Write output:
+        if num_timestep % pvd_period == 0:
+            # compute radially averaged temperature profile
+            averager.extrapolate_layer_average(T_avg, averager.get_layer_average(T))
+            # compute deviation from layer average
+            T_dev.assign(T-T_avg)
+            paraview_file.write(u, p, T, T_dev)
+
+        if num_timestep % dumping_period == 0:
+            with CheckpointFile("./simulation_states.h5", mode="a") as chkpoint_file:
+                chkpoint_file.save_function(
+                    T,
+                    name="Temperature",
+                    idx=num_timestep,
+                    timestepping_info={"time": time, "delta_t": delta_t}
+                )
+
+        # Log diagnostics:
+        plog.log_str(f"{num_timestep} {time} {float(dt)} "
+                     f"{gd.u_rms()} {gd.T_avg()}")
+    plog.close()
+
+
+def generate_mesh():
+    import os
+    if os.path.isfile("./simulation_state.h5"):
+        return
+
+    # Set up geometry:
+    ref_level, nlayers = 7, 64
+
+    # Variable radial resolution
+    # Initiating layer heights with 1.
+    resolution_func = np.ones((nlayers))
+
+    # A gaussian shaped function
+    def gaussian(center, c, a):
+        return a * np.exp(
+            -((np.linspace(rmin, rmax, nlayers) - center) ** 2) / (2 * c**2)
+        )
+
+    # building the resolution function
+    for idx, r_0 in enumerate([rmin, rmax, rmax - 660 / 6370]):
+        # gaussian radius
+        c = 0.15
+        # how different is the high res area from low res
+        res_amplifier = 5.0
+        resolution_func *= 1 / (1 + gaussian(center=r_0, c=c, a=res_amplifier))
+
+    # Construct a CubedSphere mesh and then extrude into a sphere - note that unlike cylindrical case, popping is     done internally here:
+    mesh2d = CubedSphereMesh(rmin, refinement_level=ref_level, degree=2)
+    mesh = ExtrudedMesh(
+        mesh2d,
+        layers=nlayers,
+        layer_height=(rmax - rmin) * resolution_func / np.sum(resolution_func),
+        extrusion_type="radial",
+    )
+
+    with CheckpointFile("simulation_states.h5", mode="w") as fi:
+        fi.save_mesh(mesh)
+
+
+def T_initialise(T, average):
+    import scipy
+    import math
```

Comment:
Generally, I don't like putting imports within methods, especially when they'll always be called. It makes it harder to see the dependencies of the file.

Reply:
Agreed! The only motivation was that `T_initialise` itself is called only when initialising. But now moved to beginning.